crossbeam-utils/src/thread.rs

```diff
@@ -116,17 +114,19 @@ pub fn scope<'env, F, R>(f: F) -> thread::Result<R>
 where
     F: FnOnce(&Scope<'env>) -> R,
 {
+    let (tx, rx) = mpsc::channel();
     let scope = Scope {
         handles: SharedVec::default(),
-        wg: WaitGroup::new(),
+        chan: tx,
```

Comment:
We're reusing mpsc here because wait group is now in `crossbeam`, right?  Generally speaking, putting things in `crossbeam` disallows sub-crates to depend each other...

Reply:
Yeah :( We need to figure out how to solve this problem, but I'd rather do that later than block the new release on the decision.

Reply:
How about the other way?  Specifically, how about putting miscellaneous things in `crossbeam-utils` rather than `crossbeam`, and then continuously migrating subsets of `crossbeam-utils` into their own sub-crate?

Reply:
That strategy is antithetical to backwards compatibility.

If you break off a subset of `crossbeam-utils`, then you break compatibility with its latest version.

However, if you break off a subset of `crossbeam`, then you don't break compatibility with its latest version. The subset goes into a new crate and you reexport it into `crossbeam` so its API stays the same.

Reply:
I would agree with both - we can't have subcrates depend on `crossbeam`, but we also don't want  to break subcrates with every release. Perhaps an unstable crate for all experimental/miscelaneous stuff could exist, like `crossbeam-utils` but with no backwards compatibility guarnatees? Then other subcrates can depend on it and if it breaks, we can modify them in the same merge/PR, thus not breaking anything (except the experimental crate of course). Does this make any sense?

Reply:
In my view, `crossbeam` already is the unstable/experimental crate. It's totally OK to break compatibility in `crossbeam` because most parts of it (its subcrates) will stay the same between consecutive versions. They usually won't change in the dependency graph.

On the other hand, smaller crates are those that are expected to be more stable. We should break off parts of `crossbeam` into dedicated crates when they mature.

The problem, of course, is that `crossbeam` is big and we can't depend on the whole `crossbeam` within smaller subcrates. That would create cycles anyway.

Personally, I think we should just shrug off minor annoyances like the one that came up here. Replacing `WaitGroup` with `std::sync::mpsc` is not the end of the world. There were several similar instances where we wanted a small crate to depend on bigger one, but it's never been something that isn't trivial to get around.

Yes, it's an eye-sore, but I just think it's such a non-issue that any possible remedy would be worse than the problem itself. :)

Reply:
Yeah, it's not about `WaitGroup`: it's a minor thing.  The discussion is on the crate organization.

It seems `crossbeam-utils` currently serves two roles: (1) collection of *solid* synchronization libraries, and (2) collection of experimental features.  This PR tries to split it and move experimental features into the main crate.

I'd like to propose moving experimental features into a new **unpublished** crate, say `crossbeam-nursery`, whose stability is not a concern.  The main crate will simply re-export features in the nursery, and other crates will use the nursery.  By doing so we can achieve both relative stability of `crossbeam-utils` and availability of experimental features by separating the two roles of `crossbeam-utils` into two crates.

Reply:
> It seems `crossbeam-utils` currently serves two roles: (1) collection of _solid_ synchronization libraries, and (2) collection of experimental features.

Just to clarify, this is the current state of `crossbeam-utils`:
https://docs.rs/crossbeam-utils/0.5.0/crossbeam_utils/

Note that `AtomicCell` and `sync::{ShardedLock,WaitGroup}` have never been published. So `crossbeam-utils` consists of: (1) scoped threads, (2) `CachePadded`, (3) `AtomicConsume`.

This PR won't split the latest version of `crossbeam-utils`. I'm just moving unpublished experimental stuff out of it.

> I'd like to propose moving experimental features into a new **unpublished** crate, say `crossbeam-nursery`, whose stability is not a concern.

Two problems with that:

1. If nursery is an unstable crate other crossbeam crates depend on, then every time nursery breaks compatibility, we have to re-publish all other crates in order not to end up with multiple dependencies of nursery in the graph (see #190).

2. `crossbeam-utils` wouldn't be able to depend on nursery at all because nursery would depend on `parking_lot`. If `crossbeam-utils` has a long list of dependencies, people will not use it in their libraries.